Celery (#30)
HackSoftware/Django-Styleguide: README.md

        self.assertEqual(expected, result)
```

## Celery

We use [Celery](http://www.celeryproject.org/) for the following general cases:

* Communicating with 3rd party services (sending emails, notifications, etc.)
* Offloading heavier computational tasks outside the HTTP cycle.
* Periodic tasks (using Celery beat)

We try to treat Celery as if it's just another interface to our core logic - meaning - **don't put business logic there.**

An exmaple task might look like this:

```python
from celery import shared_task

from project.app.services import some_service_name as service

> @shared_task

Reviewer: ```suggestion

@shared_task
```

def some_service_name(*args, **kwargs):
    service(*args, **kwargs)
```
This is a task, having the same name as a service, which holds the actual business logic.

> **Of course, we can have more complex cituations**, like a chain or chord of tasks, each of them doing different domain relateed logic. In that case, it's hard to isolate everything in a service, because we now have dependencies between the tasks.

Reviewer: ```suggestion
**Of course, we can have more complex situations**, like a chain or chord of tasks, each of them doing different domain related logic. In that case, it's hard to isolate everything in a service, because we now have dependencies between the tasks.
```

If that happens, we try to expose an interface to our domain & let the tasks work with that interface.

One can argue that having an ORM object is an interface by itself, and that's true. Sometimes, you can just update your object from a task & that's OK.

But there are times where you need to be strict and don't let tasks do database calls straight from the ORM, but rather, via an exposed interface for that.

**More complex scenarios depend on their context. Make sure you are aware of the architecture & the decisions you are making.**

### Structure

#### Configuration

> We put Celery confguration in a Django app called `tasks`. The [Celery config](https://docs.celeryproject.org/en/latest/django/first-steps-with-django.html) itself is located in `apps.py`, in `TasksConfig.ready` method.

Reviewer: ```suggestion
We put Celery configuration in a Django app called `tasks`. The [Celery config](https://docs.celeryproject.org/en/latest/django/first-steps-with-django.html) itself is located in `apps.py`, in `TasksConfig.ready` method.
```

This Django app also holds any additional utilities, related to Celery.

Here's an example `project/tasks/apps.py` file:

```python
import os
from celery import Celery
from django.apps import apps, AppConfig

> from django.conf import settings

Reviewer: Maybe split imports as in other code snippets?

Something like:

```python
import os

from celery import Celery

from django.apps import apps, AppConfig
from django.conf import settings
```

Author: Agreed

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'config.settings.local')

app = Celery('project')

class TasksConfig(AppConfig):
    name = 'project.tasks'
    verbose_name = 'Celery Config'

    def ready(self):
        app.config_from_object('django.conf:settings', namespace="CELERY")
        app.autodiscover_tasks()

@app.task(bind=True)
def debug_task(self):
    from celery.utils.log import base_logger
    base_logger = base_logger

    base_logger.debug('debug message')
    base_logger.info('info message')
    base_logger.warning('warning message')
    base_logger.error('error message')
    base_logger.critical('critical message')

    print('Request: {0!r}'.format(self.request))

    return 42
```

#### Tasks

Tasks are located in in `tasks.py` modules in different apps.

We follow the same rules as with everything else (APIs, services, selectors): **if the tasks for a given app grow too big, split them by domain.**

> Meaning, you can end up with `tasks/domain_a.py` and `tasks/domain_b.py`. All you need to do is impor them in `tasks/__init__.py` for Celery to autodiscover them.

Reviewer: ```suggestion
Meaning, you can end up with `tasks/domain_a.py` and `tasks/domain_b.py`. All you need to do is import them in `tasks/__init__.py` for Celery to autodiscover them.
```

The general rule of thumb is - split your tasks in a way that'll make sense to you.

### Periodic Tasks

Managing periodic tasks is quite important, especially when you have tens, or hundreds of them.

> We use [Celery Beat](https://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) + `django_celery_beat.schedulers:DatabaseScheduler` + [`django-celery-beat`](https://github.com/celery/django-celery-beat) for our peridoic tasks.

Reviewer: ```suggestion
We use [Celery Beat](https://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) + `django_celery_beat.schedulers:DatabaseScheduler` + [`django-celery-beat`](https://github.com/celery/django-celery-beat) for our periodic tasks.
```
